controllers/nifinodegroupautoscaler_controller.go

```diff
@@ -0,0 +1,428 @@
+		clusterPatch := runtimeClient.MergeFromWithOptions(cluster.DeepCopy(), runtimeClient.MergeFromWithOptimisticLock{})
+
+		if numDesiredReplicas > numCurrentReplicas {
+			// need to increase node group
+			numNodesToAdd := numDesiredReplicas - numCurrentReplicas
+			r.Log.Info(fmt.Sprintf("Adding %d more nodes to cluster %s spec.nodes configuration for node group %s", numNodesToAdd, cluster.Name, nodeGroupAutoscaler.Spec.NodeConfigGroupId))
+
+			startingNodeId := maxId + 1
+			if err = r.scaleUp(nodeGroupAutoscaler, cluster, numNodesToAdd, startingNodeId); err != nil {
+				return RequeueWithError(r.Log, fmt.Sprintf("Failed to scale cluster %s up for node group %s", cluster.Name, nodeGroupAutoscaler.Spec.NodeConfigGroupId), err)
+			}
+
+		} else if numDesiredReplicas < numCurrentReplicas {
+			// need to decrease node group
+			numNodesToRemove := numCurrentReplicas - numDesiredReplicas
+			r.Log.Info(fmt.Sprintf("Removing %d nodes from cluster %s spec.nodes configuration for node group %s", numNodesToRemove, cluster.Name, nodeGroupAutoscaler.Spec.NodeConfigGroupId))
+
+			if err = r.scaleDown(nodeGroupAutoscaler, cluster, numNodesToRemove); err != nil {
+				return RequeueWithError(r.Log, fmt.Sprintf("Failed to scale cluster %s down for node group %s", cluster.Name, nodeGroupAutoscaler.Spec.NodeConfigGroupId), err)
+			}
+		}
+
+		// patch nificluster resource with added/removed nodes
+		if err = r.Client.Patch(ctx, cluster, clusterPatch); err != nil {
+			return RequeueWithError(r.Log, fmt.Sprintf("Failed to patch nifi cluster with changes in nodes. Tried to apply the following patch:\n %v+", clusterPatch), err)
+		}
+
+		// update autoscaler state to InSync.
+		if err = r.updateAutoscalerReplicaState(ctx, nodeGroupAutoscaler, v1alpha1.AutoscalerStateInSync); err != nil {
+			return RequeueWithError(r.Log, fmt.Sprintf("Failed to udpate node group autoscaler state for node group %s", nodeGroupAutoscaler.Spec.NodeConfigGroupId), err)
+		}
+	} else {
+		r.Log.V(5).Info("Cluster replicas config and current number of replicas are the same", "replicas", nodeGroupAutoscaler.Spec.Replicas)
+	}
+	//TODO ensure finalizers on NifiNodeGroupAutoscaler
+
+	// update replica and replica status
+	if err = r.updateAutoscalerReplicaStatus(ctx, cluster, nodeGroupAutoscaler); err != nil {
+		return RequeueWithError(r.Log, fmt.Sprintf("Failed to update node group autoscaler replica status for node group %s", nodeGroupAutoscaler.Spec.NodeConfigGroupId), err)
+	}
+
+	return reconcile.Result{
+		RequeueAfter: util.GetRequeueInterval(r.RequeueInterval, r.RequeueOffset),
+	}, nil
+}
+
+// scaleUp updates the provided cluster.Spec.Nodes list with the appropriate numNodesToAdd according to the autoscaler.Spec.UpscaleStrategy
+func (r *NifiNodeGroupAutoscalerReconciler) scaleUp(autoscaler *v1alpha1.NifiNodeGroupAutoscaler, cluster *v1alpha1.NifiCluster, numNodesToAdd int32, startingNodeId int32) error {
```

Comment:
I agree - Do you have a preference for answering the question "how do we choose new node ids"? Users can choose whatever node Ids they want for the initial nodes (e.g. 1, 2, 50, 7000), so I see a couple of options:

1. We order the node Ids in the node group and iterate until we find a gap, like you suggest in [this](https://github.com/konpyutaika/nifikop/pull/89#discussion_r872537245) comment. 

2. We always choose higher node Ids. Order the node Ids, find the maxId and every new node Id is relative to that maxId. 

With either approach, we'd need to ensure that the node Id isn't being used by another node group.

Reply:
There are pros and cons for both options, as long as we keep the logic centralized in one place, we can make an arbitrary choice without thinking too much and see how it works (it will be easy to change it later!). 

So no preference, you can choose the one you like best :D

Reply:
I elected to go with the approach you suggested (i.e. 1.) and implemented that. Let me know your thoughts!